metadata-ingestion/src/datahub/ingestion/source/redash.py

```diff
@@ -0,0 +1,466 @@
+        url = f"/api/data_sources/{data_source_id}"
+        resp = self.client._get(url).json()
+        return resp
+
+    @lru_cache(maxsize=None)
+    def _get_datasource_urn_from_data_source_id(self, data_source_id):
+        data_source = self._get_chart_data_source(data_source_id)
+        data_source_type = data_source.get("type")
+        data_source_name = data_source.get("name")
+        data_source_options = data_source.get("options", {})
+
+        if data_source_type:
+            map = REDASH_DATA_SOURCE_TO_DATAHUB_MAP.get(
+                data_source_type, {"platform": DEFAULT_DATA_SOURCE_PLATFORM}
+            )
+            platform = map.get("platform")
+            platform_urn = f"urn:li:dataPlatform:{platform}"
+
+            db_name_key = map.get("db_name_key", "db")
+            db_name = data_source_options.get(db_name_key, DEFAULT_DATA_BASE_NAME)
+
+            # Redash Query Results
+            if data_source_type == "results":
+                dataset_urn = (
+                    f"urn:li:dataset:("
+                    f"{platform_urn},{data_source_name},{self.config.env})"
+                )
+                return data_source_type, dataset_urn
+
+            # Other Redash supported data source as in REDASH_DATA_SOURCE_TO_DATAHUB_MAP
+            if db_name:
+                dataset_urn = (
+                    f"urn:li:dataset:(" f"{platform_urn},{db_name},{self.config.env})"
+                )
+                return data_source_type, dataset_urn
+        return data_source_type, None
+
+    def _get_dashboard_description_from_widgets(
+        self, dashboard_widgets: List[Dict]
+    ) -> str:
+        description = ""
+
+        for widget in dashboard_widgets:
+            visualization = widget.get("visualization")
+            if not visualization:
+                options = widget.get("options")
+                text = widget.get("text")
+                isHidden = widget.get("isHidden")
+
+                # TRICKY: If top-left most widget is a Textbox, then we assume it is the Description
+                if options and text and not isHidden:
+                    position = options.get("position")
+                    if position:
+                        col = position.get("col")
+                        row = position.get("row")
+                        if col == 0 and row == 0:
+                            description = text
+                else:
+                    continue
+
+        return description
+
+    def _get_dashboard_chart_urns_from_widgets(
+        self, dashboard_widgets: List[Dict]
+    ) -> List[str]:
+        chart_urns = []
+        for widget in dashboard_widgets:
+            # In Redash, chart is called visualization
+            visualization = widget.get("visualization")
+            if visualization:
+                visualization_id = visualization.get("id", "unknown")
+                chart_urns.append(f"urn:li:chart:({self.platform},{visualization_id})")
```

Comment:
Wouldn't this create a chart URN with unknown ID? Would it be better to skip adding unknown charts rather than using this placeholder?

Reply:
Already created handler to skip unknown chart